exp/query/lsm_fetcher.go

```diff
@@ -0,0 +1,131 @@
+	DownloadToPath(ctx context.Context, collection, tenant, nodeName, path string) error
+}
+
+// LSMCache provides caching of tenant's LSM data.
+type LSMCache interface {
+	Tenant(collection, tenantID string) (*TenantCache, error)
+	AddTenant(collection, tenantID string, version uint64) error
+	BasePath() string
+}
+
+// NewLSMFetcher creates LSMFetcher without cache.
+// Every call is forwarded to given `upstream` to fetch the LSM data for the tenant.
+func NewLSMFetcher(basePath string, upstream LSMDownloader, log logrus.FieldLogger) *LSMFetcher {
+	return NewLSMFetcherWithCache(basePath, upstream, nil, log)
+}
+
+// NewLSMFetcherWithCache creates LSMFetcher with cache.
+func NewLSMFetcherWithCache(basePath string, upstream LSMDownloader, cache LSMCache, log logrus.FieldLogger) *LSMFetcher {
+	return &LSMFetcher{
+		basePath: basePath,
+		cache:    cache,
+		log:      log,
+		upstream: upstream,
+	}
+}
+
+// Fetch try to fetch LSM data for the given (collection,tenant).
+// It works in following modes
+// 1. If cache is nil, fetch from upstream.
+// 2. If cache is enabled and not found in cache, fetch from upstream.
+// 3. If cache is enabled and found in the cache but local version is outdated, fetch from the upstream
+// 4. if cache is enabled, found in the cache and local version is up to date, return from the cache.
+func (l *LSMFetcher) Fetch(ctx context.Context, collection, tenant string, version uint64) (*lsmkv.Store, string, error) {
+	var (
+		basePath          = l.basePath
+		tenantPath string = ""
+		download   bool   = true
+	)
+
+	if l.cache != nil {
+		basePath = l.cache.BasePath()
+		c, err := l.cache.Tenant(collection, tenant)
+		if err == nil && version != 0 && c.Version >= version {
+			tenantPath = c.AbsolutePath()
+			download = false
+		}
+		if err != nil && !errors.Is(err, ErrTenantNotFound) {
+			return nil, "", err
+		}
+	}
+
+	if download {
+		tenantPath = path.Join(
+			basePath,
+			strings.ToLower(collection),
+			strings.ToLower(tenant),
+		)
+		// src - s3://<collection>/<tenant>/<node>/
+		// dst (local) - <data-path/<collection>/<tenant>
+		l.log.WithFields(logrus.Fields{
+			"collection": collection,
+			"tenant":     tenant,
+			"nodeName":   nodeName,
+			"dst":        tenantPath,
+		}).Debug("starting download to path")
+		if err := l.upstream.DownloadToPath(ctx, collection, tenant, nodeName, tenantPath); err != nil {
+			return nil, "", err
+		}
+
+	}
+
+	// populate the cache before returning
+	if l.cache != nil {
+		if err := l.cache.AddTenant(collection, tenant, version); err != nil {
+			return nil, "", fmt.Errorf("failed populate the cache: %w", err)
+		}
+	}
+
+	lsmPath := path.Join(tenantPath, defaultLSMRoot)
+
+	store, err := lsmkv.New(lsmPath, lsmPath, l.log, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
+	if err != nil {
+		return nil, "", fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
+	}
+
+	return store, lsmPath, nil
```

Comment:
Shouldn't this method return early if there is a cache hit?

The current implementation pushes a new entry onto the evictlist for every search (unbounded) and also calls `usage()` on every search.

Can you do something like?

```suggestion
	var (
		basePath          = l.basePath
		tenantPath string = ""
	)

	if l.cache != nil {
		basePath = l.cache.BasePath()
		c, err := l.cache.Tenant(collection, tenant)
		if err == nil && version != 0 && c.Version >= version {
			tenantPath = c.AbsolutePath()
			// return lsmkv.New here
	}

	tenantPath = path.Join(
		basePath,
		strings.ToLower(collection),
		strings.ToLower(tenant),
	)
	// src - s3://<collection>/<tenant>/<node>/
	// dst (local) - <data-path/<collection>/<tenant>
	l.log.WithFields(logrus.Fields{
		"collection": collection,
		"tenant":     tenant,
		"nodeName":   nodeName,
		"dst":        tenantPath,
	}).Debug("starting download to path")
	if err := l.upstream.DownloadToPath(ctx, collection, tenant, nodeName, tenantPath); err != nil {
		return nil, "", err
	}

	// populate the cache before returning
	if l.cache != nil {
		if err := l.cache.AddTenant(collection, tenant, version); err != nil {
			return nil, "", fmt.Errorf("failed populate the cache: %w", err)
		}
	}

	lsmPath := path.Join(tenantPath, defaultLSMRoot)

	store, err := lsmkv.New(lsmPath, lsmPath, l.log, nil, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop())
	if err != nil {
		return nil, "", fmt.Errorf("failed to create store to read offloaded tenant data: %w", err)
	}

	return store, lsmPath, nil
```

Reply:
A related issue we should leave a future TODO for is that the current implementation will download the same tenant for each versioned search request until the tenant version is downloaded and in the cache (eg if 10 requests come for the same tenant come at ~the same time, this will run 10 downloads from s3).

Reply:
Hmm. another good question 👍 

I actually have an idea to handle this case using Go's [singleflight](https://pkg.go.dev/golang.org/x/sync/singleflight) package. 

tldr; we create single `singleflight.Group` for each `key` of the cache. So that if 10 concurrent requests comes in for single key and if it's a cache miss, then instead of spanning out 10 requests to upstream, we do one upstream call and share the results with remaining requests (other 9 would wait meanwhile).

It's very handy package I have used it in the past. I didn't do it straightway to to keep the PR simple for now. Happy to add a TODO. good catch. Thanks.

Reply:
Also worth mentioning that this solves the multiple in-flight requests problem only in single process (say requests coming to single querier replica in our case). 

If say we have 10 in-flight requests for same tenant but being handled by multiple queriers, still each querier have to make at most one upstream call.

There are ways to avoid this as well (say using like [Groupcache)](https://github.com/golang/groupcache) but adds more complexity and definitely not worth doing unless we have some real usage data IMHO

Reply:
> If say we have 10 in-flight requests for same tenant but being handled by multiple queriers, still each querier have to make at most one upstream call.

Agree, it's ok if request lands on different queriers that queriers will all hit S3. I like the singleflight pkg I had never seen it before.

Reply:
Done. singleflight will come in follow up PR